### ISSUE TITLE
the selected build of GDB does not support Python scripting. it cannot be used in qt creator.

### ISSUE THREAD

**Issue by mumin16** [brechtsanders/winlibs_mingw#2]
the selected build of GDB does not support Python scripting.
it cannot be used in qt creator.

**Comment by mumin16** [comment]
in clion:

Error during pretty printers setup: Undefined info command: "pretty-printer".  Try "help info".

**Comment by brechtsanders** [comment]
Do you have a simple test case for this, or specific gdb build parameters that were missing?

**Comment by brechtsanders** [comment]
I see now that GDB was built without python (and also without tcl and tk).
I will try to build with them.
Right now I'm compiling the new GCC 9.3, so I will see if I can get this sorted for GDB for the next winlibs personal build release.

**Comment by brechtsanders** [comment]
I just tried with GDB included built with --with-python.
It now also contains a file `pretty_printers.py`, but I'm not sure if the path in the package should be:
 `mingw64/share/gdb/share/gdb/python/gdb/command/pretty_printers.py`
or:
 `mingw64/share/gdb/python/gdb/command/pretty_printers.py`
Do you know this?

Also, if I make a release with this gdb will be broken for the users that download it as it will fail to start because the Python DLL file is missing, and I don't intend to distribute an entire Python in the package.

**Comment by anb0s** [comment]
The second one `mingw64/share/gdb/python/gdb/command/pretty_printers.py` should be the right path.

**Comment by anb0s** [comment]
> Also, if I make a release with this gdb will be broken for the users that download it as it will fail to start because the Python DLL file is missing, and I don't intend to distribute an entire Python in the package.

You can use the [Embedded Python](https://www.python.org/ftp/python/3.8.2/python-3.8.2-embed-amd64.zip) and copy only the `python38.dll` and `python38.zip` from the zip file to `/bin`.
This is not that big ~6 MB.

I did it manually for my own GDB build and it worked. I do not now if a patch was needed, as i have no access to my PC...

**Comment by brechtsanders** [comment]
The concept behind winlibs is that *every* dependancy is build from source.
That way everything is build with the same compiler and linked with the same standard C/C++ libraries.
So far I haven't build Python from source yet, but maybe I should give it another try.

**Comment by anb0s** [comment]
As you mentioned the Python DLL is missing and must be copied by user:
![image](https://user-images.githubusercontent.com/95811/77560526-e2baf900-6ebd-11ea-925f-29c883da6c4c.png)

**Comment by anb0s** [comment]
After adding python.dll etc. GDB starts, but it misses the files, looks like path `d:\r\winlibs64-9.3.0-7.0.0\inst_gdb-9.1\share\gdb\share\gdb/python` is hardcoded?

```
d:\git\s7p\work2\s7p.tools\win64_gcc_mingw64_9.3.0\bin>gdb
Python Exception <class 'ModuleNotFoundError'> No module named 'gdb':
d:\git\s7p\work2\s7p.tools\win64_gcc_mingw64_9.3.0\bin\gdb.exe: warning:
Could not load the Python gdb module from `d:\r\winlibs64-9.3.0-7.0.0\inst_gdb-9.1\share\gdb\share\gdb/python'.
Limited Python support is available from the _gdb module.
Suggest passing --data-directory=/path/to/gdb/data-directory.
GNU gdb (GDB for MinGW-W64 x86_64, built by Brecht Sanders) 9.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-w64-mingw32".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb)
```

**Comment by anb0s** [comment]
It also be good if it can be compiled with Python 3.7 (python3.7.dll) or 3.8 (python3.8.dll)...

**Comment by brechtsanders** [comment]
I haven't tested this at all.
It is very well possible that compiling this in uses absolute paths.
For now I will not put any more effort in this as Python or CPython won't build yet in my environment, and the main focus of winlibs is compiled languages like C and C++ and the goal is to build every component in the same environment.

Once I get CPython to compile I may reavaluate.
Currently I got 3.8.2 to compile, but linking fails with some undefined references.

**Comment by anb0s** [comment]
Thank You for the effort! If you have a new version i would be happy to test it :)

**Comment by nenin-sc** [comment]
Any news about python support?

**Comment by Trass3r** [comment]
I wonder why they require python support in QtCreator in the first place.

**Comment by krsch** [comment]
Is it possible to help with this issue? I can find no build or packaging scripts so I don't know how to experiment with solving this issue.

**Comment by Guyutongxue** [comment]
> Is it possible to help with this issue? I can find no build or packaging scripts so I don't know how to experiment with solving this issue.

https://github.com/niXman/mingw-builds Thid repo provides build scripts with Python support. I've tried these scripts, and built MinGW with GCC 10.2.0 successfully.

**Comment by hamidb80** [comment]
is it possible to configure GDB with external python?

**Comment by longnguyen2004** [comment]
@brechtsanders I'm currently in the process of adding Python support to the [llvm-mingw](https://github.com/mstorsjo/llvm-mingw) toolchain. If you wanna add support for Python in your toolchain, check out the [MSYS2 patchset for python3](https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-python). It should allow you to build Python from source. Just apply all the patches, run `autoreconf -fi` and `configure` like usual. Have a look at the PKGBUILD file in the patchset for the configure options.

**Comment by emiro85** [comment]
Any update on this issue? I would love to move from MinGW builds to your package, but missing Python support is a blocker.

**Comment by brechtsanders** [comment]
Still unable to build python/cpython in my environment :-(

I need to find time to go through the patches in the links you sent...

**Comment by longnguyen2004** [comment]
You can have a look at our Python build script [here](https://github.com/mstorsjo/llvm-mingw/blob/python/build-python.sh)
In short, we basically build a native `cpython`, then we build `libffi` and `cpython` for mingw-w64, using the patchset from MSYS2. After that, we removed a few stuff to reduce the size of the package (namely the `test` module and the bytecode cache).

**Comment by hamidb80** [comment]
try to use this version of minGW https://nim-lang.org/install_windows.html

**Comment by brechtsanders** [comment]
@hamidb80 What do you mean exactly?
https://winlibs.com/ provides a standalone build of the latest MinGW-w64 version (currently 9.0.0).
The download on the link you mention seems to be an older winlibs release (MinGW-w64 8.0.2).

**Comment by hamidb80** [comment]
> @hamidb80 What do you mean exactly?
> https://winlibs.com/ provides a standalone build of the latest MinGW-w64 version (currently 9.0.0).
> The download on the link you mention seems to be an older winlibs release (MinGW-w64 8.0.2).

it supports GDB scripting [I think that's what you wanted, right? ]

**Comment by brechtsanders** [comment]
@hamidb80 No, this question is specifically about Python support in GDB.
Since the https://nim-lang.org/install_windows.html is an earlier download from https://winlibs.com/ that was renamed, it doesn't have this either.

**Comment by hamidb80** [comment]
> @hamidb80 No, this question is specifically about Python support in GDB.
> Since the https://nim-lang.org/install_windows.html is an earlier download from https://winlibs.com/ that was renamed, it doesn't have this either.

They have a debugger that is based on gdb scripting with python

**Comment by xtemp09** [comment]
Guyutongxue made a perfect bundle with [a complete mingw-w64 toolchain](https://github.com/Guyutongxue/mingw-release/releases). It works perfectly with Qt Creator.

**Comment by xtemp09** [comment]
Any news on python support? LLVM 12.0.1 has been released.

Using your toolchain, I compiled clang with default targetting libc++ instead of libstdc++, using Win32 threads instead of pthreads and libc++abi instead of libgcc_s (the compilation process contains a flaw, libc++abi prefers pthreads, it needs to be set off explicitly).

With Guyutongxue's bundle I made clangless gcc and gccless clang (clang defaults to compiler-rt).

**Comment by xtemp09** [comment]
Will python scripting in gdb & lldb be implemented before New Year's day?

**Comment by brechtsanders** [comment]
No. Python still won't build due the winlibs way. But from time to time I try to get it to build (last time was last week when Python 3.11.0a1 came out).

**Comment by longnguyen2004** [comment]
Don't use 3.10.0 or above, the MSYS2 team only supports up to 3.9.7 (3.10.0 is ongoing). Use the pre-patched Python from https://github.com/msys2-contrib/cpython-mingw to build for mingw

**Comment by brechtsanders** [comment]
@longnguyen2004 Thanks for the tip. Unfortunately the cpython-mingw sources won't build out of the box either and there is no documentation about how to build it :-(

**Comment by longnguyen2004** [comment]
Here's what I do for building Python:
Assuming `$PREFIX` is the build prefix, `$PREFIX_NATIVE` is the prefix for native tools (if cross-compiling), `$HOST` is the host for the build

1. Cloning libffi and python:
```sh
git clone https://github.com/libffi/libffi -b v3.4.2
git clone https://github.com/python/cpython -b v3.9.7 # If cross-compiling
git clone https://github.com/msys2-contrib/cpython-mingw -b mingw-v3.9.7
# This is a branch, not a tag. For reproducibility, use a tag here

cd libffi && autoreconf -vfi && cd ..
cd cpython-mingw && autoreconf -vfi && cd ..
```

2. (if cross-compiling) Build native libffi and cpython
```sh
cd libffi
mkdir build-native && cd build-native
../configure --prefix="$PREFIX_NATIVE" --disable-symvers --disable-docs
make install -j8

cd ../../cpython
mkdir build-native && cd build-native
../configure --prefix="$PREFIX_NATIVE" --without-ensurepip --with-system-ffi \
    CFLAGS="-I$PREFIX_NATIVE/include"   \
    CXXFLAGS="-I$PREFIX_NATIVE/include" \
    LDFLAGS="-L$PREFIX_NATIVE/lib"
make install -j8

cd ../..
# You can reuse this native build as long as the native version matches the cross version
```

3. Build mingw libffi and cpython
```sh
export PATH="$PREFIX_NATIVE/bin:$PATH" # If cross-compiling

cd libffi
mkdir build-$HOST && cd build-$HOST
../configure --prefix="$PREFIX" --host=$HOST --disable-symvers --disable-docs
make install -j8

cd ../../cpython-mingw
mkdir build-$HOST && cd build-$HOST
BUILD=$(../config.guess)
export CC=$HOST-gcc  # $HOST-clang
export CXX=$HOST-g++ # $HOST-clang++
../configure --prefix="$PREFIX" --build=$BUILD --host=$HOST \
    CFLAGS="-I$PREFIX/include" CXXFLAGS="-I$PREFIX/include" LDFLAGS="-L$PREFIX/lib" \
    --enable-shared             \
    --with-system-ffi           \
    --without-ensurepip         \
    --without-c-locale-coercion
make install -j8
```

Then build gdb with `--with-python="$PREFIX`

**Comment by brechtsanders** [comment]
Doesn't work out of the box.
First error I get is ´Python/dynload_shlib.c:87:34: error: implicit declaration of function 'dlsym' [-Werror=implicit-function-declaration]´

**Comment by longnguyen2004** [comment]
Are you cross-compiling or not? Also, please give me the command invocation and the stdout.

**Comment by brechtsanders** [comment]
Native build on Windows in MSYS2 shell but using winlibs compiler and all prerequisites already compiled from source in the same environment.

I just noticed I was building 3.9.5 as that was the last tag visible in https://github.com/msys2-contrib/cpython-mingw/tags and since my build scripts use tarballs instead of git clones.

**Comment by brechtsanders** [comment]
I figured stuff out now.

It's till a mystery to me why the tarballs on https://github.com/msys2-contrib/cpython-mingw/tags don't work and why 3.9.7 isn't even listed in there.

Anyway, it looks like you will still get gdb and lldb with python before X-mas.

Is there a minimal case for testing (I myself have never debugged Python like that)?

**Comment by emiro85** [comment]
Thats awesome news. Its not about debugging Python, but using Python scripts to enhance the gdb debugging experience. The Qt project and its IDE QtCreator are using it heavily. Hence, if you want to debug using QtCreator, gdb needs to be built with Python support. A minimal test case would be to install QtCreator, create a simple C++ application, compile and try to debug it through QtCreator.

**Comment by longnguyen2004** [comment]
> It's till a mystery to me why the tarballs on https://github.com/msys2-contrib/cpython-mingw/tags don't work and why 3.9.7 isn't even listed in there.

Because the tags are from the upstream repo. The MSYS2 team doesn't create tags for their fork, it's an ongoing patchset that's based on a specific Python version. My "use a tag" advice is a bit misleading here. I actually mean "use a commit hash". Sorry for that.

> Is there a minimal case for testing
The bare minimum is being able to launch Python in GDB (by running `python` from the gdb prompt) and import the `gdb` module. If that succeeds, then the basic functionality is there, and we can solve other problems later.

**Comment by xtemp09** [comment]
Perhaps, a minimal test case is the possibility to be used in _Qt Creator_ as the debugger.

Arcane has been released, my birthday is coming, now Python support is finally being added to the toolchain. The end is nigh.

**Comment by longnguyen2004** [comment]
> Perhaps, a minimal test case is the possibility to be used in _Qt Creator_ as the debugger.
> 
> Arcane has been released, my birthday is coming, now Python support is finally being added to the toolchain. The end is nigh.

Only after a year :)
(which is kinda funny, since [llvm-mingw](https://github.com/mstorsjo/llvm-mingw) also took a year to add Python support, and I basically wrote the build script for them)

**Comment by brechtsanders** [comment]
![image](https://user-images.githubusercontent.com/17645177/145815101-f1084df4-19d6-456e-81dc-62be160240b4.png)

**Comment by longnguyen2004** [comment]
Can you import `gdb`?

**Comment by brechtsanders** [comment]
What do you mean exactly?

I have started building GCC 11.2.1-snapshot20211211 and will try include gdb in this package.

Do you know if it only requires the Python DLL, or should other files also be copied (e.g. `lib/python*/*`)?

**Comment by longnguyen2004** [comment]
> What do you mean exactly?

After entering Python, run `import gdb` to import the `gdb` module. If it succeed, then the basic functionality is working.

> Do you know if it only requires the Python DLL, or should other files also be copied (e.g. `lib/python*/*`)?

I just do `make install` and leave it as is, since the user might want to do custom data processing that relies on the standard modules.
To lower the build size, you can delete the various test modules and the `__pycache__` folder, to save a couple hundred MB.

**Comment by brechtsanders** [comment]
I'm trying to make the distribution package.

Currently when I run `gdb --version` the result is:
```
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Python path configuration:
  PYTHONHOME = (not set)
  PYTHONPATH = (not set)
  program name = 'r:\d\prog\winlibs64ucrt_stage\custombuilt\share\python/bin/python'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = 'r:\\d\\prog\\winlibs64ucrt_stage\\custombuilt\\share\\python/bin/python'
  sys.base_prefix = 'R:/winlibs64ucrt_stage/33664.cpython-mingw/inst_cpython-mingw-3.9.7/share/python'
  sys.base_exec_prefix = 'R:/winlibs64ucrt_stage/33664.cpython-mingw/inst_cpython-mingw-3.9.7/share/python'
  sys.platlibdir = 'lib'
  sys.executable = 'r:\\d\\prog\\winlibs64ucrt_stage\\custombuilt\\share\\python/bin/python'
  sys.prefix = 'R:/winlibs64ucrt_stage/33664.cpython-mingw/inst_cpython-mingw-3.9.7/share/python'
  sys.exec_prefix = 'R:/winlibs64ucrt_stage/33664.cpython-mingw/inst_cpython-mingw-3.9.7/share/python'
  sys.path = [
    'R:/winlibs64ucrt_stage/33664.cpython-mingw/inst_cpython-mingw-3.9.7/share/python/lib/python39.zip',
    'R:/winlibs64ucrt_stage/33664.cpython-mingw/inst_cpython-mingw-3.9.7/share/python/lib/python3.9',
    'R:/winlibs64ucrt_stage/33664.cpython-mingw/inst_cpython-mingw-3.9.7/share/python/lib/lib-dynload',
    '',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'

Current thread 0x000033e4 (most recent call first):
<no Python frame>
```

Looks like the executable path isn't properly detected and there are some dive letter issues (e.g. `r:\\d\\`).

**Comment by brechtsanders** [comment]
This fixed the issue. Apparently the detected path got overruled by the compile-time value...
```
patch -ulbf Python/pathconfig.c << EOF
@@ -421,7 +421,7 @@
             } \
         }

-#ifdef MS_WINDOWS
+#if defined(MS_WINDOWS) && !defined(__MINGW32__)
     if (config->executable != NULL && config->base_executable == NULL) {
         /* If executable is set explicitly in the configuration,
            ignore calculated base_executable: _PyConfig_InitPathConfig()
EOF
```

Now the next step is to run: `python -c "import gdb"` but unfortunately this returns:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "R:/winlibs64ucrt_stage/inst_winlibs-x86_64-posix-seh-gcc-11.2.1-snapshot20211211-mingw-w64ucrt-9.0.0-r1/mingw64/lib/python3.9/gdb/__init__.py", line 19, in <module>
    import _gdb
ModuleNotFoundError: No module named '_gdb'
```
What needs to be copied for the module with underscore (`_gdb`) to be able to load?

**Comment by longnguyen2004** [comment]
You need to run it from inside gdb, not from bare Python.
Basically, run gdb, then enter Python mode, then `import gdb`

**Comment by brechtsanders** [comment]
Like this?
```
GNU gdb (GDB for MinGW-W64 x86_64, built by Brecht Sanders) 11.1
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-w64-mingw32".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) python-interactive
>>> import gdb
>>> quit()
```

Looks good right?

**Comment by Whotakesmyname** [comment]
Looks beautiful.

**Comment by brechtsanders** [comment]
Please try this release: https://github.com/brechtsanders/winlibs_mingw/releases/tag/11.2.1-snapshot20211211-9.0.0-ucrt-r1

Can anybody test this with Qt Creator?

**Comment by longnguyen2004** [comment]
Haven't tested with Qt Creator, but Code::Blocks and VS Code work fine.
Since gcc provides pretty printers for libstdc++, we can configure gdb to load it automatically.
Assuming `$PREFIX` is the build prefix, `$GCC_VERSION` is the used gcc version.

1. Build gdb with the `--with-system-gdbinit` flag:
`./configure --prefix="$PREFIX" --with-system-gdbinit="$PREFIX/share/gdb/gdbinit" ...`
2. After building gcc and gdb, write into `$PREFIX/share/gdb/gdbinit`:
```sh
echo << EOF > $PREFIX/share/gdb/gdbinit
python
import sys
import os
sys.path.append(os.path.join(sys.prefix, "share", "gcc-$GCC_VERSION", "python"))
from libstdcxx.v6.printers import register_libstdcxx_printers
register_libstdcxx_printers(None)
end
EOF
```

This is the output of `info pretty-printer` after registering:
<details>
    <summary>Pretty printers</summary>

```
(gdb) info pretty-printer
global pretty-printers:
  builtin
    mpx_bound128
  libstdc++-v6
    __gnu_cxx::_Slist_iterator
    __gnu_cxx::__8::_Slist_iterator
    __gnu_cxx::__8::__normal_iterator
    __gnu_cxx::__8::slist
    __gnu_cxx::__normal_iterator
    __gnu_cxx::slist
    __gnu_debug::_Safe_iterator
    std::_Bit_const_iterator
    std::_Bit_iterator
    std::_Bit_reference
    std::_Deque_const_iterator
    std::_Deque_iterator
    std::_Fwd_list_const_iterator
    std::_Fwd_list_iterator
    std::_List_const_iterator
    std::_List_iterator
    std::_Node_handle
    std::_Rb_tree_const_iterator
    std::_Rb_tree_iterator
    std::__8::_Bit_const_iterator
    std::__8::_Bit_iterator
    std::__8::_Bit_reference
    std::__8::_Deque_const_iterator
    std::__8::_Deque_iterator
    std::__8::_Fwd_list_const_iterator
    std::__8::_Fwd_list_iterator
    std::__8::_List_const_iterator
    std::__8::_List_iterator
    std::__8::_Node_handle
    std::__8::_Rb_tree_const_iterator
    std::__8::_Rb_tree_iterator
    std::__8::__cxx11::__cxx1998::list
    std::__8::__cxx11::basic_string
    std::__8::__cxx11::list
    std::__8::__cxx1998::_Bit_const_iterator
    std::__8::__cxx1998::_Bit_iterator
    std::__8::__cxx1998::_Bit_reference
    std::__8::__cxx1998::_Deque_const_iterator
    std::__8::__cxx1998::_Deque_iterator
    std::__8::__cxx1998::_Fwd_list_const_iterator
    std::__8::__cxx1998::_Fwd_list_iterator
    std::__8::__cxx1998::_List_const_iterator
    std::__8::__cxx1998::_List_iterator
    std::__8::__cxx1998::bitset
    std::__8::__cxx1998::deque
    std::__8::__cxx1998::forward_list
    std::__8::__cxx1998::list
    std::__8::__cxx1998::map
    std::__8::__cxx1998::multimap
    std::__8::__cxx1998::multiset
    std::__8::__cxx1998::set
    std::__8::__cxx1998::unordered_map
    std::__8::__cxx1998::unordered_multimap
    std::__8::__cxx1998::unordered_multiset
    std::__8::__cxx1998::unordered_set
    std::__8::__cxx1998::vector
    std::__8::any
    std::__8::basic_string
    std::__8::basic_string_view
    std::__8::bitset
    std::__8::deque
    std::__8::error_code
    std::__8::error_condition
    std::__8::experimental::filesystem::v1::__cxx11::path
    std::__8::experimental::filesystem::v1::path
    std::__8::experimental::fundamentals_v1::any
    std::__8::experimental::fundamentals_v1::basic_string_view
    std::__8::experimental::fundamentals_v1::optional
    std::__8::filesystem::__cxx11::path
    std::__8::filesystem::path
    std::__8::forward_list
    std::__8::list
    std::__8::map
    std::__8::multimap
    std::__8::multiset
    std::__8::optional
    std::__8::pair
    std::__8::partial_ordering
    std::__8::priority_queue
    std::__8::queue
    std::__8::set
    std::__8::shared_ptr
    std::__8::stack
    std::__8::strong_ordering
    std::__8::tr1::shared_ptr
    std::__8::tr1::unordered_map
    std::__8::tr1::unordered_multimap
    std::__8::tr1::unordered_multiset
    std::__8::tr1::unordered_set
    std::__8::tr1::weak_ptr
    std::__8::tuple
    std::__8::unique_ptr
    std::__8::unordered_map
    std::__8::unordered_multimap
    std::__8::unordered_multiset
    std::__8::unordered_set
    std::__8::variant
    std::__8::vector
    std::__8::weak_ordering
    std::__8::weak_ptr
    std::__cxx11::__cxx1998::list
    std::__cxx11::basic_string
    std::__cxx11::list
    std::__cxx1998::_Bit_const_iterator
    std::__cxx1998::_Bit_iterator
    std::__cxx1998::_Bit_reference
    std::__cxx1998::_Deque_const_iterator
    std::__cxx1998::_Deque_iterator
    std::__cxx1998::_Fwd_list_const_iterator
    std::__cxx1998::_Fwd_list_iterator
    std::__cxx1998::_List_const_iterator
    std::__cxx1998::_List_iterator
    std::__cxx1998::bitset
    std::__cxx1998::deque
    std::__cxx1998::forward_list
    std::__cxx1998::list
    std::__cxx1998::map
    std::__cxx1998::multimap
    std::__cxx1998::multiset
    std::__cxx1998::set
    std::__cxx1998::unordered_map
    std::__cxx1998::unordered_multimap
    std::__cxx1998::unordered_multiset
    std::__cxx1998::unordered_set
    std::__cxx1998::vector
    std::__debug::bitset
    std::__debug::deque
    std::__debug::forward_list
    std::__debug::list
    std::__debug::map
    std::__debug::multimap
    std::__debug::multiset
    std::__debug::priority_queue
    std::__debug::queue
    std::__debug::set
    std::__debug::stack
    std::__debug::unique_ptr
    std::__debug::unordered_map
    std::__debug::unordered_multimap
    std::__debug::unordered_multiset
    std::__debug::unordered_set
    std::__debug::vector
    std::any
    std::basic_string
    std::basic_string_view
    std::bitset
    std::deque
    std::error_code
    std::error_condition
    std::experimental::filesystem::v1::__cxx11::path
    std::experimental::filesystem::v1::path
    std::experimental::fundamentals_v1::any
    std::experimental::fundamentals_v1::basic_string_view
    std::experimental::fundamentals_v1::optional
    std::filesystem::__cxx11::path
    std::filesystem::path
    std::forward_list
    std::list
    std::map
    std::multimap
    std::multiset
    std::optional
    std::pair
    std::partial_ordering
    std::priority_queue
    std::queue
    std::set
    std::shared_ptr
    std::stack
    std::strong_ordering
    std::tr1::shared_ptr
    std::tr1::unordered_map
    std::tr1::unordered_multimap
    std::tr1::unordered_multiset
    std::tr1::unordered_set
    std::tr1::weak_ptr
    std::tuple
    std::unique_ptr
    std::unordered_map
    std::unordered_multimap
    std::unordered_multiset
    std::unordered_set
    std::variant
    std::vector
    std::weak_ordering
    std::weak_ptr
```
</details>

**Comment by brechtsanders** [comment]
I want all my build to be relocatable. The `--with-system-gdbinit=` flag you propose is an absolute path. Is there any way to make it relative to gdb.exe?

**Comment by longnguyen2004** [comment]
If the path to `--with-system-gdbinit` contains the build prefix, then it's automatically relocatable.

**Comment by emiro85** [comment]
> Please try this release: https://github.com/brechtsanders/winlibs_mingw/releases/tag/11.2.1-snapshot20211211-9.0.0-ucrt-r1
> 
> Can anybody test this with Qt Creator?

Tested with Qt Creator and works perfectly!

**Comment by xtemp09** [comment]
Happy birthday to me! Perhaps, it's time to close this issue.

I have three questions loosely related to the topic:

- [Why are all the dates 2019-11-11](https://github.com/brechtsanders/winlibs_mingw/tags)? What happened on this date, besides the greatest singles' day on AliExpress?


- [Why are releases sorted not by release date](https://github.com/brechtsanders/winlibs_mingw/releases)? The first release is nine days ago, next ― a few hours ago, then nine days ago again.


- Brecht, what do you use to automate the process? Perhaps, you should publish the build_all script on github.

I can dream that one day, I just start the script in busybox on VM and it automagically downloads necessary files, builds everything from the ground, producing the final 7z-archive of the toolchain.

I automated build process for gcc and clang; though, of course, LLVM developers break build system every minor release and introduce new workaround-related bugs, rendering build_all script null and void.

If you publish the script(s), here, on github, people will take a look at it. They can fix collectively the build system-related troubles faster. It's been two years mumin16 opened this issue, longnguyen2004 solved it, however, it remained unknown to you.

**Comment by brechtsanders** [comment]
I just do releases, I don't do anything with specifically tags, but apparently a tag is created for each release with the date the repository was created...

I guess GitHub sorts by name, not by date. I had a GCC 12 snapshot release and then more recent GCC 11 stuff which is sorted after GCC 12. Why? That's a question for GitHub...

The process is automated using my own tools, which I have recently published at https://github.com/brechtsanders/winlibs_tools . The next step will be also publish the build recipes used by these tools. They are not just for GCC, but over 2000 projects! Please be patient, it's coming...

The tools are native Windows (for now), so you won't need a busybox VM, just a real Windows system. But the tools do build on Linux too, just the recipes are all still native (for now).

**Comment by Platzer** [comment]
First: I wanted to thank @brechtsanders for maintaining this project 🎉

Second @longnguyen2004 :
> Haven't tested with Qt Creator, but Code::Blocks and VS Code work fine. Since gcc provides pretty printers for libstdc++, we can configure gdb to load it automatically. Assuming `$PREFIX` is the build prefix, `$GCC_VERSION` is the used gcc version.

I downloaded the "winlibs-x86_64-posix-seh-gcc-11.2.1-snapshot20211211-mingw-w64ucrt-9.0.0-r1" snapshot and tried it with VS Code and it didn't work. My "info pretty-printer" output is:
```
(gdb) info pretty-printer
global pretty-printers:
  builtin
    mpx_bound128
```
Do I've to configure something in gdb first?

**Comment by longnguyen2004** [comment]
> Do I've to configure something in gdb first?

This build doesn't have the autoloading code yet, so you have to load the pretty printers manually. I'm working with him to add that in and get it out in (hopefully) the next snapshot build.

**Comment by brechtsanders** [comment]
I will look into the `--with-system-gdbinit=` option for building `gdb` as soon as I get a properly working MSYS2 environment again.

**Comment by longnguyen2004** [comment]
Try submitting an issue at https://github.com/msys2/msys2-runtime?

**Comment by brechtsanders** [comment]
@longnguyen2004 Thanks! Done (https://github.com/msys2/msys2-runtime/issues/75)

**Comment by brechtsanders** [comment]
Just released https://github.com/brechtsanders/winlibs_mingw/releases/tag/11.2.0-9.0.0-msvcrt-r4 wit the proposed gdbinit script. Can somebody please test this?

**Comment by Whotakesmyname** [comment]
> Just released https://github.com/brechtsanders/winlibs_mingw/releases/tag/11.2.0-9.0.0-msvcrt-r4 wit the proposed gdbinit script. Can somebody please test this?

@brechtsanders Thanks for hardworking these days.

Wrote a simple test trying to pretty print a STL vector `a = {1, 2, 3, 4, 5}`, the new pre-release's result:
```
(gdb) p a
$1 = {<std::_Vector_base<int, std::allocator<int> >> = {
    _M_impl = {<std::allocator<int>> = {<__gnu_cxx::new_allocator<int>> = {<No data fields>}, <No data fields>}, <std::_Vector_base<int, std::allocator<int> >::_Vector_impl_data> = {_M_start = 0x29b8f851940, _M_finish = 0x29b8f851954,
        _M_end_of_storage = 0x29b8f851954}, <No data fields>}}, <No data fields>}
```

For reference, an old but work version from Guyutongxue mentioned above gave an output:
```
(gdb) p a
$1 = std::vector of length 5, capacity 5 = {1, 2, 3, 4, 5}
```

---

I noticed that the pre-release's gdb configurations about Python support is not relocatable. That might be the problem:
```
~#@❯ ./gdb --configuration                                                                                       ❮  
This GDB was configured as follows:
   configure --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32
             --with-auto-load-dir=$debugdir:$datadir/auto-load
             --with-auto-load-safe-path=$debugdir:$datadir/auto-load
             --with-expat
             --with-gdb-datadir=/R/winlibs64_stage/inst_gdb-11.1/share/gdb/share/gdb (relocatable)
             --with-jit-reader-dir=/R/winlibs64_stage/inst_gdb-11.1/share/gdb/lib/gdb (relocatable)
             --without-libunwind-ia64
             --with-lzma
             --without-babeltrace
             --without-intel-pt
             --with-mpfr
             --with-xxhash
             --with-python=D:/Prog/winlibs64_stage/custombuilt/share/python
             --with-python-libdir=/d/Prog/winlibs64_stage/custombuilt/share/python/lib
             --without-debuginfod
             --without-guile
             --enable-source-highlight
             --with-separate-debug-dir=/R/winlibs64_stage/inst_gdb-11.1/share/gdb/lib/debug (relocatable)
             --with-system-gdbinit=/R/winlibs64_stage/inst_gdb-11.1/share/gdb/gdbinit (relocatable)

("Relocatable" means the directory can be moved with the GDB installation
tree, and GDB will still find it.)
```

For reference, the configurations from Guyutongxue's version:
```
~#@❯ gdb --configuration                                                                             ❮  6s 202ms   
C:\tools\mingw64\bin\gdborig.exe: warning: Couldn't determine a path for the index cache directory.
This GDB was configured as follows:
   configure --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32
             --with-auto-load-dir=$debugdir:$datadir/auto-load
             --with-auto-load-safe-path=$debugdir:$datadir/auto-load
             --with-expat
             --with-gdb-datadir=/build/x86_64-1110-posix-seh-rt_v7/mingw64/share/gdb (relocatable)
             --with-jit-reader-dir=/build/x86_64-1110-posix-seh-rt_v7/mingw64/lib/gdb (relocatable)
             --without-libunwind-ia64
             --without-lzma
             --without-babeltrace
                             --without-intel-pt
             --with-mpfr
             --without-xxhash
             --with-python=/build/x86_64-1110-posix-seh-rt_v7/mingw64/opt (relocatable)
             --without-guile
             --disable-source-highlight
             --with-separate-debug-dir=/build/x86_64-1110-posix-seh-rt_v7/mingw64/lib/debug (relocatable)
             --with-system-gdbinit=/build/x86_64-1110-posix-seh-rt_v7/mingw64/etc/gdbinit (relocatable)

("Relocatable" means the directory can be moved with the GDB installation
tree, and GDB will still find it.)
```

I think we are close!

**Comment by longnguyen2004** [comment]
It doesn't matter if the Python dir is relocatable or not, since Python does load correctly.

Firstly, the gdbinit file doesn't exist (@brechtsanders did you forget to add my code snippet above?)
Secondly, seems like the path passed to `--with-system-gdbinit` is slightly incorrect, which causes gdb to look at the wrong directory.

`gdb --configuration` returns this
```
This GDB was configured as follows:
   configure --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32
             --with-auto-load-dir=$debugdir:$datadir/auto-load
             --with-auto-load-safe-path=$debugdir:$datadir/auto-load
             --with-expat
             --with-gdb-datadir=/R/winlibs64_stage/inst_gdb-11.1/share/gdb/share/gdb (relocatable)
             --with-jit-reader-dir=/R/winlibs64_stage/inst_gdb-11.1/share/gdb/lib/gdb (relocatable)
             --without-libunwind-ia64
             --with-lzma
             --without-babeltrace
             --without-intel-pt
             --with-mpfr
             --with-xxhash
             --with-python=D:/Prog/winlibs64_stage/custombuilt/share/python
             --with-python-libdir=/d/Prog/winlibs64_stage/custombuilt/share/python/lib
             --without-debuginfod
             --without-guile
             --enable-source-highlight
             --with-separate-debug-dir=/R/winlibs64_stage/inst_gdb-11.1/share/gdb/lib/debug (relocatable)
             --with-system-gdbinit=/R/winlibs64_stage/inst_gdb-11.1/share/gdb/gdbinit (relocatable)
```

`show data-directory` returns this:
```
(gdb) show data-directory
GDB's data directory is "c:\gcc-msvcrt\share\gdb".
```

Cross-checking the 2 outputs, the build prefix seems to be `/R/winlibs64_stage/inst_gdb-11.1/share/gdb/`, so the system gdbinit flag translates to `--with-system-gdbinit=$prefix/gdbinit`, which is wrong.

@Whotakesmyname Try dropping this file into the root of the toolchain, and rename it to `gdbinit`, like so
![image](https://user-images.githubusercontent.com/15729831/146581371-d9634f6b-28e8-43d9-9df2-1c1975548143.png)
[gdbinit.txt](https://github.com/brechtsanders/winlibs_mingw/files/7736628/gdbinit.txt)

@brechtsanders Can you share the configure command for gdb? I need to double check the prefix stuff.

**Comment by brechtsanders** [comment]
Yes, I really want everything in my distribution to be relocatable.

Unfortunately I made a mistake with the `--with-system-gdbinit` path in this first build, sorry.

The new release should find `gdbinit` in the right place.

Get it while it's hot: https://github.com/brechtsanders/winlibs_mingw/releases/tag/11.2.0-9.0.0-msvcrt-r5

**Comment by longnguyen2004** [comment]
`--with-system-gdbinit` looks correct now, we just need to create the gdbinit file

Add the following to the end of your gcc build
```sh
echo << EOF > $PREFIX/share/gdb/gdbinit
python
import sys
import os
sys.path.insert(0, os.path.join(sys.prefix, "share", "gcc-$GCC_VERSION", "python"))
from libstdcxx.v6.printers import register_libstdcxx_printers
register_libstdcxx_printers(None)
end
EOF
```
(replace the variable names with the respective ones in your build script)

Here's how it looks like on my side, after expansion
[gdbinit.txt](https://github.com/brechtsanders/winlibs_mingw/files/7738559/gdbinit.txt)

**Comment by brechtsanders** [comment]
Sorry, that was a mistake in my package script (guess it was Friday night...).

Repackaged version released here: https://github.com/brechtsanders/winlibs_mingw/releases/tag/11.2.0-9.0.0-msvcrt-r4c

**Comment by brechtsanders** [comment]
> @brechtsanders Thanks for hardworking these days.

You're welcome. I'm glad you guys all helped making the winlibs build more widely useable.

**Comment by longnguyen2004** [comment]
![image](https://user-images.githubusercontent.com/15729831/146635801-80c4ea8a-9251-4583-a45b-f92a0671fcc9.png)
Are you getting enough sleep? That file name shouldn't be there in the release notes :)

The good news tho, pretty printers are registered successfully
```
GNU gdb (GDB for MinGW-W64 x86_64, built by Brecht Sanders) 11.1
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-w64-mingw32".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) info pretty-printer
global pretty-printers:
  builtin
    mpx_bound128
  libstdc++-v6
    __gnu_cxx::_Slist_iterator
    __gnu_cxx::__8::_Slist_iterator
    __gnu_cxx::__8::__normal_iterator
    __gnu_cxx::__8::slist
    __gnu_cxx::__normal_iterator
    __gnu_cxx::slist
    __gnu_debug::_Safe_iterator
    std::_Bit_const_iterator
    std::_Bit_iterator
    std::_Bit_reference
    std::_Deque_const_iterator
    std::_Deque_iterator
    std::_Fwd_list_const_iterator
    std::_Fwd_list_iterator
    std::_List_const_iterator
    std::_List_iterator
    std::_Node_handle
    std::_Rb_tree_const_iterator
    std::_Rb_tree_iterator
    std::__8::_Bit_const_iterator
    std::__8::_Bit_iterator
    std::__8::_Bit_reference
    std::__8::_Deque_const_iterator
    std::__8::_Deque_iterator
    std::__8::_Fwd_list_const_iterator
--Type <RET> for more, q to quit, c to continue without paging--
```

Anyway, good work everyone! I'm waiting for the UCRT build, so I can finally have 2 fully featured toolchains with Python support.

**Comment by brechtsanders** [comment]
@longnguyen2004 No, definitely not sleeping well these days. May dayjob is very busy with the Apache Log4j Security Vulnerabilities [CVE-2021-45046](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-45046) and [CVE-2021-44228](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-44228). I was never a fan of Java, and that's not going to change...

**Comment by Whotakesmyname** [comment]
> @Whotakesmyname Try dropping this file into the root of the toolchain, and rename it to `gdbinit`, like so

@longnguyen2004  Yes, though I did not put that file into that location but after manually type those settings into the running gdb session, it did work.

---

> I was never a fan of Java

@brechtsanders Then we share something in common :)

**Comment by longnguyen2004** [comment]
Thank god I chose C# over Java.
Anyway this is pretty much solved, so we just have to wait for others to test.

**Comment by brechtsanders** [comment]
> I'm waiting for the UCRT build, so I can finally have 2 fully featured toolchains with Python support.

Busy building UCRT with gdb with python...

![image](https://user-images.githubusercontent.com/17645177/146648509-b47d8dc5-4632-475e-8763-bf48289c81a7.png)

BTW: I just tried to build Qt Creator with winlibs gcc and it builds just fine. Would it be interesting to have a complete Qt environment in one package? If yes, what would need to be in it (besides GCC toolchain, Qt and Qt Creator)?

**Comment by longnguyen2004** [comment]
I don't think we need to go that far. Qt is a huge library, and most people who are using Qt Creator are either working on CMake projects (so the Qt library would be of no use), or they are already building Qt from source. It's not worth a few hundred MB of extra compressed data (which can results in a few extra GB uncompressed).

For me, this is good enough, as it checked many of my criteria for a good toolchain (UCRT support, recent compilers, Python support for IDE integration).

**Comment by xtemp09** [comment]
I'm a Qt fan, though, I always compile two builds ― static and shared, choosing appropriate depending on the case.

Compilation of Qt is fast and simple in comparison to compilation of LLVM and gcc, so it won't really change anything.

brechtsanders, you can produce special _[stable]_ builds containing Qt and boost. Perhaps, it's too tedious to make these builds all the time.

**Comment by Whotakesmyname** [comment]
> Would it be interesting to have a complete Qt environment in one package?

I would say it is interesting to have some other binaries, but at least not in one package. They can be distributed in different packages, or even different projects. It is not part of mingw, but kind of dependent on mingw.

Personally, I would expect this package as a Windows mingw version of build-essentials.

**Comment by xtemp09** [comment]
I agree, it's a good idea, to have separate packages of Qt.

**Comment by brechtsanders** [comment]
UCRT release: https://github.com/brechtsanders/winlibs_mingw/releases/tag/11.2.0-9.0.0-ucrt-r3

**Comment by xtemp09** [comment]
Dear @brechtsanders, please, don't consider I cavil about trifles or point out mistakes.
1)	Have you considered a possibility to include python into a separate directory? A separate directory for python of the toolchain will be distinguished from the python already present. If I add `mingw64/bin` to `%PATH%`, I will have two python directories, since I already have one python installed.

@Guyutongxue places python into `mingw64/opt/bin`, while `mingw64/bin` has only gdb.

2)	Why do you not use symbolic links? `mingw64/x86_64-w64-mingw32/bin` has some binaries that coincide with that of `mingw64/bin`. `mingw64/bin` also contains identical files as well, for example, `c++.exe`, `g++.exe`, `x86_64-w64-mingw32-c++.exe`, `x86_64-w64-mingw32-g++.exe` or `clang.exe`, `clang-cl.exe`, `clang-cpp.exe` and `clang++.exe`. 7z preserves symbolic links. Attached is the file I used after compilation of gcc; you might want to look at it ([create_links.txt](https://github.com/brechtsanders/winlibs_mingw/files/7740853/create_links.txt)). I used hardlinks on my system.

3) What do you think of [this](https://github.com/meganz/mingw-std-threads)? This repository allows usage of std::thread and [this one](https://github.com/josuttis/jthread) allows usage of std::jthread without libwinpthread.

**Comment by brechtsanders** [comment]
1. If the Python DLL (`libpython3.9.dll`) is not in the same path as `gdb` there would be errors when launching `gdb`.
  I could leave `python*.exe` out of the package or move it to a separate folder, but there are no guarantees `libpython3.9.dll` won't conflict with your other Python distribution.
2. I avoid symbolic links at all times on Windows because under some circumstances people may not have rights to use symbolic links on Windows, and I really want this package to be portable (in the sense of PortableApps).
3. `mingw-std-threads` was created for something lacking on classic MinGW, but in MinGW-w64 this is no longer missing (please correct me if I'm wrong). `jthread` builds just fine and people can use it if they like - I see no reason to include it in the compiler tools distribution.

**Comment by Whotakesmyname** [comment]
> Have you considered a possibility to include python into a separate directory?  ... , I will have two python directories, since I already have one python installed.

  ~~Now this bundle is all you need and you should delete other Pythons.~~ Yes that would be more considerate but for a workaround, I think you can just delete those Python binaries because they are not used. GDB links Python inside.

---

mingw-std-threads I'm not sure including this in a GNU bundle is a good idea. Since this is a header library, maybe it should be left to programmers who are interested to include it in their projects with or without a cmake file.

**Comment by brechtsanders** [comment]
> GDB configuration paths regarding Python are not relocatable.

I'm not so sure about that. Those configure flags are for building GDB. Python itself seems to be quite relocatable, at least after I patched `Python/initconfig.c` to search relative to the Python DLL (I will publish the build recipes in due time).

**Comment by Whotakesmyname** [comment]
Yes, I didn't see any problems caused by that (since GDB links Python instead of executing it) so far. If you are comfortable with your absolute path of your system appears in distributions, I think that's fine.

**Comment by longnguyen2004** [comment]
1. This can be done by setting `--with-python-libdir` to a subfolder inside the prefix (Guyutongxue's build: `$prefix/opt/lib`, llvm-mingw: `$prefix/python/lib`), however I don't think this is the right solution here. As @brechtsanders have pointed out, you would still have a `libpython3.x.dll` potentially conflicting with other Python installations. If you don't want the local Python to have priority, just place it below/after the global Python in `PATH`
2. Also the same as Brecht
3. That library looks unmaintained to me, so I'd recommend against using it. The current implementation of libstdc++ `thread` combined with `libwinpthread` is good enough for most people. Makes me wonder why no one bothers to make a proper `thread` impl for libstdc++...

> the GDB configuration paths regarding Python are not relocatable.

Doesn't matter. After reading through gdb's source, and from my experience adding Python support to llvm-mingw, it seems like specifying `--with-python-libdir=$prefix/lib` or whatever is enough, and [gdb will automatically relocate it](https://github.com/bminor/binutils-gdb/blob/5ec501b5749c1df79b749a7fd885760dc734b31c/gdb/main.c#L725-L728) and help Python find the proper libdir. Maybe you don't even need to patch `Python/initconfig.c` after all.

**Comment by xtemp09** [comment]
@longnguyen2004, Guyutongxue puts all python stuff in `mingw64/opt/bin`, including `libpython.dll`, `libcrypto.dll` and `libssl.dll`. Therefore, `mingw64/bin` is completely pythonless and can't cause potential conflicts.

All right, what about Qt and Qt Creator? @brechtsanders, are you planning on building separate Qt packages or make special builds containing Qt? Official toolchains contain Qt Creator built with MSVS, even if mingw-w64's Qt is installed. Qt Creator has its internal clang, clang-tidy, clangd, clazy, jom, yaml, libcrypto, libssl and MSVC Qt dlls by default. Distribution of Qt as bundle will be much smaller than official since it won't have duplicated Qt*.dlls and clang.

**Comment by xtemp09** [comment]
@brechtsanders, I'm trying to compile Qt with the bundle you have just released. Errors occured during configuration, please take a look at the [configuration log](https://github.com/brechtsanders/winlibs_mingw/files/7742930/conf.log).

Possibly, the most important fragment is

> Configuring 'qttools'
> -- Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR) 
> -- Could NOT find LibXml2 (missing: LIBXML2_LIBRARY LIBXML2_INCLUDE_DIR) 
> CMake Error at E:/temp/mingw64/lib/cmake/llvm/LLVMExports.cmake:596 (message):
>   The imported target "LLVMDemangle" references the file
> 
>      "E:/temp/mingw64/lib/libLLVMDemangle.dll.a"
> 
>   but this file does not exist.  Possible reasons include:
> 
>   * The file was deleted, renamed, or moved to another location.
>   * An install or uninstall procedure did not complete successfully.
>   * The installation package was faulty and contained
> 
>      "E:/temp/mingw64/lib/cmake/llvm/LLVMExports.cmake"
> 
>   but not all the files it references.



> python --version
> Python 3.9.7

> perl --version
> 
> This is perl 5, version 32, subversion 1 (v5.32.1) built for MSWin32-x64-multi-thread
> 
> Copyright 1987-2021, Larry Wall
> 
> Perl may be copied only under the terms of either the Artistic License or the
> GNU General Public License, which may be found in the Perl 5 source kit.
> 
> Complete documentation for Perl, including FAQ lists, should be found on
> this system using "man perl" or "perldoc perl".  If you have access to the
> Internet, point your browser at http://www.perl.org/, the Perl Home Page.

> gcc --version
> gcc (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders) 11.2.0
> Copyright (C) 2021 Free Software Foundation, Inc.
> This is free software; see the source for copying conditions.  There is NO
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

> clang --version
> (built by Brecht Sanders) clang version 13.0.0
> Target: x86_64-w64-windows-gnu
> Thread model: posix
> InstalledDir: E:\temp\mingw64\bin

>  uname -a
> Windows_NT Comp 10.0 19042 x86_64 MS/Windows

>echo $PATH
> C:/Windows/system32;C:/Windows;C:/Windows/System32/WindowsPowerShell/v1.0;c:/bin;E:/temp/mingw64/bin;C:/Qt/Toolchains/perl/bin;C:/bin/win_flex_bison-latest;C:/bin/swig;C:/Qt/Toolchains/cmake/bin

**Comment by longnguyen2004** [comment]
> `"E:/temp/mingw64/lib/libLLVMDemangle.dll.a"`

Maybe he configured LLVM with `-DLLVM_INSTALL_TOOLCHAIN_ONLY=1`, so the different libraries aren't installed.
Not sure if a full LLVM build is viable though, since it can easily add hundreds of MB into the compressed bundle.

> Therefore, mingw64/bin is completely pythonless and can't cause potential conflicts.

gytx achieved this by renaming `gdb` to `gdborig` and make a thin wrapper that adds the local Python to `PATH` and call `gdborig`. I suppose we can do the same thing with a batch wrapper
```batch
@echo off
rem Change the path below to suit the build configuration
set PYTHON_PREFIX_REL=%~dp0\..\opt\bin
pushd %PYTHON_PREFIX_REL%
set PYTHON_PREFIX_ABS=%CD%
popd
set PATH=%PYTHON_PREFIX_ABS%;%PATH%

rem Passing args to gdb
shift
set GDB_ARGS=
:loop
if [%0] == [] goto break
set GDB_ARGS=%GDB_ARGS% %0
goto loop
:break

gdborig %GDB_ARGS%
```
(not tested, use with caution)
But once again, this seems like unnecessary work to me, given that placing the toolchain path after the system Python path would work as well, and is way easier.

**Comment by xtemp09** [comment]
Possibly, I'll need QDoc. clang-LLVM 12.0.1 I compiled contains:

```
llvm/lib # du -csh libLLVM* | tail -1
31.8M   total
```

```
llvm/lib # du -csh libclang* | tail -1
31.4M   total
```

```
llvm/lib # du -sh libc++* libu* libomp*
1.9M    libc++.a
688.0K  libc++.dll.a
388.0K  libc++abi.a
12.0K   libc++experimental.a
48.0K   libunwind.a
8.0K    libunwind.dll.a
124.0K  libomp.dll.a
```

**Comment by longnguyen2004** [comment]
@brechtsanders Are you interested in making a full LLVM build? (toolchain, lib, headers). Just be aware of the potential size increase.

**Comment by brechtsanders** [comment]
Sorry, was a bit busy today.

I build LLVM without the `LLVM_INSTALL_TOOLCHAIN_ONLY` cmake flag.

I'm starting to regret putting LLVM and GCC in one distribution package. I think I should make a completely separate LLVM/CLang package, actually built against LLVM's standard library. It would also make sense as a build environment to eventually target ARM64 (aarch64).

What do you guys think?

(maybe this should be a separate ticket and we should close this one)

**Comment by Whotakesmyname** [comment]
> But once again, this seems like unnecessary work to me, given that placing the toolchain path after the system Python path would work as well, and is way easier.

Yes, the precedence of PATH env var surely works, as long as we explicitly document it, since we do not provide an install script and that needs users to manually set.

Not sure the motivation to include LLVM within GCC bundle. Agree with separating them into different packages.

**Comment by xtemp09** [comment]
Astrologers have announced a week of Log4j CVEs.

Well, this thing should be thought over.

I usually used your toolchain to compile my own, clang uses compiler-rt, libunwind and lld, while libc++ was pthreadlessly compiled as both static and shared with a static pthreadless libc++abi. I have separate directories for gcc and clang. clang and gcc share hardlinked libiconv, libintl, zlib, liblua, liblzma, libxml2 and other libraries. So, they have the same msvcrt, libs, DLLs, x86_64-w64-mingw32/lib and include.

**Comment by longnguyen2004** [comment]
Let's close this issue first, since Python has been cracked :)

**Comment by xtemp09** [comment]
Dear @brechtsanders, I'm trying to compile `Qt 6.2.2` with `winlibs-x86_64-posix-seh-gcc-11.2.0-mingw-w64ucrt-9.0.0-r3` you published a couple of posts earlier. The compilation fails with countless errors, for example:

```
E:/temp/qt-everywhere-src-6.2.2/qtmultimedia/src/multimedia/platform/windows/common/mfmetadata.cpp: In function 'void setUInt32Property(IPropertyStore*, const PROPERTYKEY&, quint32)':
E:/temp/qt-everywhere-src-6.2.2/qtmultimedia/src/multimedia/platform/windows/common/mfmetadata.cpp:330:19: error: 'InitPropVariantFromUInt32' was not declared in this scope; did you mean 'InitPropVariantFromInt64'?
  330 |     if (SUCCEEDED(InitPropVariantFromUInt32(ULONG(value), &propValue))) {
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~
E:/temp/qt-everywhere-src-6.2.2/qtmultimedia/src/multimedia/platform/windows/common/mfmetadata.cpp: In function 'void setUInt64Property(IPropertyStore*, const PROPERTYKEY&, quint64)':
E:/temp/qt-everywhere-src-6.2.2/qtmultimedia/src/multimedia/platform/windows/common/mfmetadata.cpp:340:19: error: 'InitPropVariantFromUInt64' was not declared in this scope; did you mean 'InitPropVariantFromInt64'?
  340 |     if (SUCCEEDED(InitPropVariantFromUInt64(ULONGLONG(value), &propValue))) {
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~
E:/temp/qt-everywhere-src-6.2.2/qtmultimedia/src/multimedia/platform/windows/common/mfmetadata.cpp: In function 'void setFileTimeProperty(IPropertyStore*, const PROPERTYKEY&, const FILETIME*)':
E:/temp/qt-everywhere-src-6.2.2/qtmultimedia/src/multimedia/platform/windows/common/mfmetadata.cpp:350:19: error: 'InitPropVariantFromFileTime' was not declared in this scope; did you mean 'InitPropVariantFromString'?
  350 |     if (SUCCEEDED(InitPropVariantFromFileTime(ft, &propValue))) {
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The error most likely caused by [this commit](https://github.com/qt/qtmultimedia/commit/a199c508b82644fddfbf9a2b3c7e03433e81bc5d). Have you tried to compile `Qt 6.2.2` recently?

I have a couple more questions also connected to the topic. 

Are you going to put an end to this issue? Are you going to move python into a separate directory, like @Guyutongxue, @jmeubank and @mstorsjo or do you prefer to keep python in /bin?

If you consider this done, I have another issue: LLDB in `winlibs-x86_64-posix-seh-gcc-11.2.0-llvm-13.0.0-mingw-w64ucrt-9.0.0-r3` does not work; its output:

```
> ./lldb
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ./lldb
#0 0x00007ff952550000
#1 0x00007ff9b2c371e9 LLDBSWIGPython_GetValueObjectSPFromSBValue(void*) (E:\temp\mingw64\bin\liblldb.dll+0x3771e9)
#2 0x00007ff9b2c33d61 LLDBSWIGPython_GetValueObjectSPFromSBValue(void*) (E:\temp\mingw64\bin\liblldb.dll+0x373d61)
#3 0x00007ff9b2ddbb21 lldb::SBDebugger::InitializeWithErrorHandling() (E:\temp\mingw64\bin\liblldb.dll+0x51bb21)
#4 0x00007ff6f3a4c11f (E:\temp\mingw64\bin\lldb.exe+0xc11f)
#5 0x00007ff6f3a41445 (E:\temp\mingw64\bin\lldb.exe+0x1445)
#6 0x00007ff6f3a41146 (E:\temp\mingw64\bin\lldb.exe+0x1146)
#7 0x00007ffa175b7034 (C:\Windows\System32\KERNEL32.DLL+0x17034)
#8 0x00007ffa19322651 (C:\Windows\SYSTEM32\ntdll.dll+0x52651)
```

Possibly, the error is connected to python. The backtrace is not really helpful since binaries in your toolchains are stripped. GDB gives:
```
> gdb lldb.exe
GNU gdb (GDB for MinGW-W64 x86_64, built by Brecht Sanders) 11.1
Reading symbols from lldb.exe...
(No debugging symbols found in lldb.exe)
(gdb) run
Starting program: E:\temp\mingw64\bin\lldb.exe
[New Thread 10328.0x46a8]
[New Thread 10328.0x5a8]
[New Thread 10328.0x2aa8]

Thread 1 received signal SIGSEGV, Segmentation fault.
0x00007ff951f70000 in ?? ()
```

**Comment by brechtsanders** [comment]
Strange, I have no issues building Qt 6.2.2 with this build.
The issues I had before with Qt have been fixed in this release (like https://bugreports.qt.io/browse/QTBUG-93452).
My configuration line like this:
```
PATH=$PYDIR:$PATH cmake.exe -Wno-dev -GNinja -DCMAKE_INSTALL_PREFIX:PATH=$INSTALLPREFIX/qt6 -DCMAKE_BUILD_TYPE:STRING=Release -DBUILD_SHARED_LIBS:BOOL=OFF -DCMAKE_C_COMPILER:FILEPATH=$(which gcc.exe) -DCMAKE_CXX_COMPILER:FILEPATH=$(which g++.exe) -DLLVM_DIR:PATH=NOTFOUND -DClang_DIR:PATH=NOTFOUND -DFEATURE_pkg_config:BOOL=ON -DFEATURE_qml_python:BOOL=ON -DFEATURE_mimetype_database:BOOL=OFF -DFEATURE_zstd:BOOL=ON -DFEATURE_egl:BOOL=OFF -DFEATURE_syslog:BOOL=OFF -DFEATURE_slog2:BOOL=OFF -DFEATURE_icu:BOOL=ON -DFEATURE_opengl:BOOL=ON -DFEATURE_opengl_desktop:BOOL=OFF -DFEATURE_opengl_dynamic:BOOL=ON -DFEATURE_system_libb2:BOOL=ON -DFEATURE_system_doubleconversion:BOOL=ON -DFEATURE_system_freetype:BOOL=ON -DFEATURE_system_harfbuzz:BOOL=ON -DFEATURE_system_jpeg:BOOL=ON -DFEATURE_system_pcre2:BOOL=ON -DFEATURE_system_png:BOOL=ON -DFEATURE_system_proxies:BOOL=ON -DFEATURE_system_zlib:BOOL=ON -DFEATURE_libproxy:BOOL=ON -DFEATURE_wmf:BOOL=OFF -DFEATURE_avfoundation:BOOL=OFF -DFEATURE_gstreamer:BOOL=ON -DFEATURE_qt3d_system_assimp:BOOL=ON -DFEATURE_system_sqlite:BOOL=ON -DFEATURE_sql_ibase:BOOL=OFF -DFEATURE_libudev:BOOL=OFF -DFEATURE_mtdev:BOOL=OFF -DFEATURE_evdev:BOOL=OFF -DBUILD_qtwayland:BOOL=OFF -DQT_USE_BUNDLED_BundledFreetype:BOOL=OFF -DQT_USE_BUNDLED_BundledHarfbuzz:BOOL=OFF -DQT_USE_BUNDLED_BundledLibpng:BOOL=OFF -DQT_USE_BUNDLED_BundledPcre2:BOOL=OFF -DFEATURE_qtplugininfo:BOOL=OFF -DBUILD_qtdoc:BOOL=OFF -DBUILD_qtquick3d:BOOL=OFF -DBUILD_qtquicktimeline:BOOL=OFF -DBUILD_qttranslations:BOOL=OFF -DBUILD_WITH_PCH:BOOL=OFF -DFEATURE_clang:BOOL=OFF -DFEATURE_clangcpp:BOOL=OFF -DTEST_libclang:BOOL=OFF -DQT_BUILD_EXAMPLES:BOOL=OFF -DQT_BUILD_EXAMPLES_BY_DEFAULT:BOOL=OFF -DQT_BUILD_BENCHMARKS:BOOL=OFF -DQT_BUILD_TESTS:BOOL=OFF -DQT_BUILD_TESTS_BY_DEFAULT:BOOL=OFF -DMySQL_INCLUDE_DIRS:PATH=$MINGWPREFIX/include/mariadb -DPYTHON_EXECUTABLE:FILEPATH=$PYDIR/python.exe -DQT_QML_PYTHON_PATH:FILEPATH=$PYDIR/python.exe -S. -Bbuild_static 
```
where `PYDIR` points to a separate Python installation.

**Comment by xtemp09** [comment]
The thing is:
```
-DFEATURE_wmf:BOOL=OFF
-DFEATURE_avfoundation:BOOL=OFF
-DFEATURE_gstreamer:BOOL=ON
```
you explicitly turned off WMF. Strange that you turned on GStreamer.

**Comment by brechtsanders** [comment]
Now that you say so, I recall I had to turn it off because it had issues, and I don't think I ever got that to work in Qt5 either.

**Comment by xtemp09** [comment]
[wiki.qt.io says](https://wiki.qt.io/Qt_6.2_Known_Issues):

> ### Qt Multimedia
> Due to these massive changes, the module does still have rough edges and probably quite a few bugs in the implementation. Reported bugs will be fixed as quickly as we can in upcoming patch level releases.
> 
> Qt Multimedia is not supported on MinGW in Qt 6.2.0. This is due to lack of WMF headers in MinGW 8.2. Work is ongoing to upgrade to MinGW 9 and with that get support for Qt Multimedia on MinGW in Qt 6.2.2.

**Comment by longnguyen2004** [comment]
@brechtsanders Can you give me the configure options for gdb and lldb? I'm thinking that the crash is related to the patch you made for Python, since we build Python for llvm-mingw and everything works fine.

**Comment by brechtsanders** [comment]
Currently I configure gdb like this:
```
./configure --prefix=$INSTALLPREFIX/share/gdb --build=$BUILDPLATFORM --host=$RUNPLATFORM --with-pkgversion="GDB for MinGW-W64 $(if ( echo $RUNPLATFORM | grep -q x86_64 ); then echo x86_64; else echo i686; fi), built by Brecht Sanders" --enable-serial-configure --disable-bootstrap --with-{mpc,mpfr,gmp,isl}=$MINGWPREFIX --disable-{nls,win32-registry,rpath,gdbtk,gtk,unit-tests} --enable-{64-bit-bfd,gdbserver,tui} --enable-64-bit-bfd --with-{expat,libiconv,zlib,lzma,curses,xxhash} --with-{tcl,tk}=$MINGWPREFIX/bin --with-system-gdbinit="$INSTALLPREFIX/share/gdb/share/gdb/gdbinit" --with-system-zlib --without-system-readline --without-{guile,x} --with-python=$MINGWPREFIX/share/python/bin/python3.exe --with-python-libdir=$MINGWPREFIX/share/python/lib --enable-compressed-debug-sections=all --enable-host-shared --with-lzma --with-included-regex --enable-source-highlight --with-mmap CFLAGS="-I$MINGWPREFIX/include/mman-win32" CXXFLAGS="-I$MINGWPREFIX/include/mman-win32" LDFLAGS="-Wl,-no-undefined -L$MINGWPREFIX/share/python/lib" LIBS="-Wl,--as-needed -lmman"
```
For lldb I'm still figuring it out as I'm now trying to build it with clang...

**Comment by xtemp09** [comment]
[From here](https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/):
> Starting with Windows 10 Insiders build 14972, symlinks can be created without needing to elevate the console as administrator.

Brecht, which Windows versions do your toolchains support? I think it should be added to winlibs.com main page.

**Comment by brechtsanders** [comment]
Windows 7 and higher for now, that's why I avoid symlinks for the time being...

**Comment by xtemp09** [comment]
Hello again, @brechtsanders, I checked out the [last release](https://github.com/brechtsanders/winlibs_mingw/releases/tag/11.2.0-14.0.0-9.0.0-ucrt-r7) and found out that `python.exe`, `python3.9.exe` and `python3w.exe` are absent. What has happened? Did you choose not not provide the binaries, only `libpython3.9.dll`?

**Comment by xtemp09** [comment]
Since it's somehow related, I'll write here. I'm trying to compile again `Qt 6.2.4` and, of course, it fails to build `qtmultimedia`, since the toolchain lacks some GUIDs in `mfapi.h`. Long time ago Liu Hao added [those GUIDs](https://sourceforge.net/p/mingw-w64/mingw-w64/ci/7fb7c9f6e401e2bc4393e61a27a68ebccffa1195/).

Qt 6 is compilable with [this toolchain](https://github.com/cristianadam/mingw-builds/releases/tag/v11.2.0). Could you update mingw-w64-headers in order to make Qt 6 compilable with your toolchain? It means to add mingw-w64-headers from the trunk.

**Comment by brechtsanders** [comment]
I prefer to use latest release builds for everything. I home MinGW-w64 gets a new release soon though.

How I build Qt 6.2.4 van be seen here: https://github.com/brechtsanders/winlibs_recipes/blob/main/recipes/qt6.winlib

I really would like to have webengine and WebKit support too. Would you happen to know how to get that to built?

**Comment by brechtsanders** [comment]
The Python .exe files were removed to avoid conflicts with people's already install Python installations (which often have specific modules installed). It's not really my goal to include a full Python version. It was only added to have Python support in the debugger, which really only needs the library.

Thanks for the tips on WebKit/WebEngine. It's a shame it's really restricted to MSVC.
I thought the Microsoft Edge browser also used the same engine and it is cross-platform, but maybe just not cross-compiler...

Maybe I should look into an experimental winlibs build with mingw-w64 trunk, but I don't think I will find the the time in the next week.

**Comment by xtemp09** [comment]
I have successsfully compiled Qt6 with `winlibs-x86_64-posix-seh-gcc-11.2.0-mingw-w64ucrt-10.0.0-r1` and llvm-windres from another toolchain, since windres from binutils has limit of 8192 bytes in the command-line (response file won't help).
I haven't tested it yet in production; the project using Qt Multimedia I wrote for Qt5, I need some time to rewrite it for Qt6. Nevertheless, everything seems to be fine.

@brechtsanders, can I request a feature - adding [mold](https://github.com/rui314/mold) and lld to the toolchain?

**Comment by brechtsanders** [comment]
That's good news.

Strange, I don't run into that windres issue in MSYS2 shell, though I haven't tried to build Qt6 with Qt Multimedia yet.

I had mold on my radar before, but it wouldn't build yet. Are you aware of successful native builds of mold?

As for lld, are you referring to LLVM's linker?
I really want to go for separate GCC and LLVM toolchains in the future, each using their respective standard libraries.

**Comment by xtemp09** [comment]
Perhaps, MSYS2 shell does not have cmd or busybox limitations, so it will work fine.

I've only read about mold, I don't know its implementations.

@brechtsanders, when should we expect separate toolchains?

**Comment by brechtsanders** [comment]
From what I read [here](https://githubhot.com/repo/rui314/mold/issues/52) [mold](https://github.com/rui314/mold) doesn't cupport COFF/PE format yet (the format used by Windows for .exe and .dll files). So it won't be possible for the same reason the gold linker (which only links ELF format) isn't possible.

If I can get LLVM/CLang to build without issues using its own standard libraries I will try to build a complete package, but so far I haven't been successful. But if you really need it you should check out https://github.com/mstorsjo/llvm-mingw

**Comment by brechtsanders** [comment]
Python support in GDB has been included for some time now
